### ISSUE TITLE
Basic auth and/or ssh url support for cargo aggregate

### ISSUE THREAD

**Issue by agraven** [mozilla/cargo-vet#336]
Using cargo-vet to aggregate audit sources from private repositories can be quite difficult, as there's no way of accessing an audit file from a private source without doing something like putting basic auth information in the URL itself, which will then result in the username:password pair being spread all over in the form of `aggregated-from` keys.

The main approaches to addressing this use case I see are:
1. A command line flag (or two) for specifying basic auth credentials. (Or maybe they could be read from stdin?)
2. Support fetching audit files via ssh, which would delegate the authentication to the ssh-agent

What your thoughts on this? I'd be happy to give writing a PR for this functionality a go

**Comment by bholley** [comment]
SSH sounds a little cleaner to me, though I'm not sure whether this would involve pulling in a bunch of dependencies or if we can lean on what cargo already does. @mystor WDYT?

**Comment by mystor** [comment]
I don't believe we currently depend on libssh2, which would probably be the easiest way to fetch something with ssh. From a quick check locally this would pull in: `ssh2`, `libssh2-sys`, `lock_api`, `openssl-sys`, `parking_lot`, `parking_lot_core`, and `scopeguard`. 

We could also probably do something without these dependencies if we're fine with just invoking `scp` on the command line and using temporary files.

**Comment by bholley** [comment]
Generally speaking I'd rather not take all those dependencies. I think I'm fine with either a command-line scp solution or an HTTP basic auth solution.

**Comment by dexterduck** [comment]
In the interim, how about adding support for the `file://` URL scheme? 

As far as I can tell it's not possible to point `cargo vet aggregate` or the `imports` table in `config.toml` to a local file (please let me know if I'm missing something), which would make it easier to create an external workaround for pulling from private repositories.

I'd be happy to put a PR in if there's interest in the change.

**Comment by mystor** [comment]
Support just for file URIs would likely be quite difficult to use, as those URIs are absolute, which isn't great for paths in a repository due to it not being cloned in a consistent location. The solution for local paths would probably need to support relative paths, which would involve tracking things like base URIs for what resources we're loading, which is a larger change.

I think the easiest way forward here is probably a PR adding support for the `ssh://` protocol by delegating to `scp`. This would require adding support to the methods in https://github.com/mozilla/cargo-vet/blob/56ebfded2d27d6c28a56832aaae1dc99b41c4311/src/network.rs.

**Comment by bachp** [comment]
I am also interested in being able to use `cargo vet aggregate` with non public repos. This would allow us to create a public supply-chain repo that is aggregated from several internal ones.

SSH would work but I could also imagine a solution where the fetcher is using the [`.netrc`](https://everything.curl.dev/usingcurl/netrc.html) file like curl.

**Comment by bachp** [comment]
@mystor Would the solution to add authrization headers as proposed in https://github.com/mozilla/cargo-vet/pull/640 be acceptable as a solution?

**Comment by mystor** [comment]
Sorry for the slow response. I've added some comments to #640, but the TL;DR is that I think something like a `--netrc-file` argument (similar to what you mention curl using in https://github.com/mozilla/cargo-vet/issues/336#issuecomment-2304754594) is probably a more appropriate option as it is more precise.
